src/app/helpers/paginationTemplate/paginationTemplate.ts

```diff
@@ -0,0 +1,198 @@
+  /**
+   * Tracks whether to display the pagination buttons
+   */
+  public displayPagination: boolean;
+  /**
+   * Tracks whether an error has occurred
+   */
+  public error: ApiErrorDetails;
+  /**
+   * Tracks whether an api request is in process
+   */
+  public loading: boolean;
+  /**
+   * Tracks additional filter constraints
+   */
+  public filter: string;
+  /**
+   * Maximum number of elements for current filter
+   */
+  public collectionSize: number;
+  /**
+   * Tracks the current filter page
+   */
+  private _page: number;
+
+  constructor(
+    protected router: Router,
+    protected route: ActivatedRoute,
+    /**
+     * Config for pagination defaults
+     */
+    protected config: NgbPaginationConfig,
+    /**
+     * API Service which will create filter requests
+     */
+    protected api: ApiFilter<M, any[]>,
+    /**
+     * Key to match filter inputs against
+     */
+    private filterKey: keyof I,
+    /**
+     * API Service parameters required to make a filter request
+     */
+    private apiParams: () => any[],
+    /**
+     * Function which will run after each api update (ie. scroll/filter)
+     */
+    private apiUpdate: (models: M[]) => void,
+    /**
+     * Default filter values, may be overridden by later requests
+     */
+    private defaultFilter: InnerFilter<I> = {}
+  ) {
+    super();
+  }
+
+  public ngOnInit() {
+    // Set pagination defaults
+    this.config.maxSize = 3;
+    this.config.pageSize = defaultApiPageSize;
+    this.config.rotate = true;
+    this.displayPagination = false;
+
+    this._page = 1;
+
+    this.apiRequest$
+      .pipe(
+        mergeMap(() => this.getModels()),
+        takeUntil(this.unsubscribe)
+      )
+      .subscribe(noop, (error: ApiErrorDetails) => {
+        console.error("Scroll Template Failure: ", error);
+        this.error = error;
+        this.loading = false;
+      });
+
+    this.updateFromUrl();
+
+    this.router.events
+      .pipe(
+        filter((event) => event instanceof NavigationEnd),
+        takeUntil(this.unsubscribe)
+      )
+      .subscribe(() => this.updateFromUrl(), noop);
+  }
+
+  /**
+   * Read the previous filter query and page number from the url,
+   * and update the page to match
+   */
+  private updateFromUrl() {
+    const params = this.route.snapshot.queryParams;
+    this.filter = params[queryKey] ?? "";
+    this._page = parseInt(params[pageKey], 10) || 1;
+    this.onFilter(this.filter, this._page);
+  }
+
+  public get page() {
+    return this._page;
+  }
+
+  /**
+   * Handle pagination events
+   */
+  public set page(page: number) {
+    this._page = page;
+    this.loading = true;
+    this.apiRequest$.next();
+    this.updateQueryParams(this._page, this.filter);
+  }
+
+  /**
+   * Handle filter events
+   */
+  public onFilter(input: string, page: number = 1) {
+    this._page = page;
+    this.filter = input;
+    this.loading = true;
+    this.apiRequest$.next();
```

Comment:
everywhere, where next is called, why not pass page and filter into the `next` function to make it immutable?